### ISSUE TITLE
M2.2.6 : BadRequest400Exception on a fresh install (index_already_exists_exception)

### ISSUE THREAD

**Issue by kanduvisla** [Smile-SA/elasticsuite#1086]
When installing Magento 2.2.6 using the CLI (it's a fresh install), I get a `BadRequest400Exception` when a `installData` script is trying to create stores.

It gave no problems in 2.2.5 and elasticsuite 2.5.x

### Preconditions

Magento Version: 2.2.6
ElasticSuite Version: 2.6.1
Environment: Developer
Elasticsearch: 5.6.8 with icu and phonetic plugins

### Steps to reproduce

Create an installData script that creates stores. Example script:

```
<?php

namespace Vendor\Module\Setup;

use Magento\Framework\Event\ManagerInterface;
use Magento\Framework\Setup\InstallDataInterface;
use Magento\Framework\Setup\ModuleContextInterface;
use Magento\Framework\Setup\ModuleDataSetupInterface;
use Magento\Store\Api\Data\StoreInterfaceFactory;
use Magento\Store\Api\GroupRepositoryInterface;
use Magento\Store\Api\StoreRepositoryInterface;
use Magento\Store\Api\WebsiteRepositoryInterface;
use Magento\Store\Model\ResourceModel\Group;
use Magento\Store\Model\ResourceModel\Store;
use Magento\Store\Model\ResourceModel\Website;

/**
 * Class InstallData
 */
class InstallData implements InstallDataInterface
{
    /**
     * @var StoreRepositoryInterface
     */
    protected $storeRepository;

    /**
     * @var Store
     */
    protected $storeResource;

    /**
     * @var StoreInterfaceFactory
     */
    protected $storeFactory;

    /**
     * @var ManagerInterface
     */
    protected $eventManager;

    /**
     * @var WebsiteRepositoryInterface
     */
    protected $websiteRepository;

    /**
     * @var Website
     */
    protected $websiteResource;

    /**
     * @var GroupRepositoryInterface
     */
    protected $groupRepository;

    /**
     * @var Group
     */
    protected $groupResource;

    /**
     * InstallData constructor.
     * @param StoreRepositoryInterface $storeRepository
     * @param Store $storeResource
     * @param StoreInterfaceFactory $storeFactory
     * @param ManagerInterface $eventManager
     * @param WebsiteRepositoryInterface $websiteRepository
     * @param Website $websiteResource
     * @param GroupRepositoryInterface $groupRepository
     * @param Group $groupResource
     */
    public function __construct(
        StoreRepositoryInterface $storeRepository,
        Store $storeResource,
        StoreInterfaceFactory $storeFactory,
        ManagerInterface $eventManager,
        WebsiteRepositoryInterface $websiteRepository,
        Website $websiteResource,
        GroupRepositoryInterface $groupRepository,
        Group $groupResource
    ) {
        $this->storeRepository = $storeRepository;
        $this->storeResource = $storeResource;
        $this->storeFactory = $storeFactory;
        $this->eventManager = $eventManager;
        $this->websiteRepository = $websiteRepository;
        $this->websiteResource = $websiteResource;
        $this->groupRepository = $groupRepository;
        $this->groupResource = $groupResource;
    }

    /**
     * @param ModuleDataSetupInterface $setup
     * @param ModuleContextInterface $context
     */
    public function install(ModuleDataSetupInterface $setup, ModuleContextInterface $context)
    {
        $this->setupWebsite();
        $this->setupStoreGroup();
        $this->setupStoreViews();
    }

    /**
     * Setup website
     */
    protected function setupWebsite()
    {
        $website = $this->websiteRepository->get('base');
        $website->setName('Website');
        $this->websiteResource->save($website);
    }

    /**
     * Setup store group
     */
    protected function setupStoreGroup()
    {
        $group = $this->groupRepository->get(1);
        $group->setName('Webshop');
        $this->groupResource->save($group);
    }

    /**
     * Setup store views
     */
    protected function setupStoreViews()
    {
        $defaultStore = $this->storeRepository->getById(\Magento\Store\Model\Store::DISTRO_STORE_ID);
        $defaultStore->setCode('nl');
        $defaultStore->setName('Nederlands');
        $this->storeResource->save($defaultStore);

        $stores = [
            ['code' => 'de', 'name' => 'Deutsch', 'sort_order' => 1],
            ['code' => 'en', 'name' => 'English', 'sort_order' => 2],
            ['code' => 'fr', 'name' => 'Français', 'sort_order' => 3]
        ];

        foreach ($stores as $storeData) {
            /** @var \Magento\Store\Model\Store $store */
            $store = $this->storeFactory->create();
            $store->setName($storeData['name']);
            $store->setCode($storeData['code']);
            $store->setSortOrder($storeData['sort_order']);
            $store->setWebsiteId($defaultStore->getWebsiteId());
            $store->setIsActive(1);
            $store->setStoreGroupId($defaultStore->getStoreGroupId());
            $this->storeResource->save($store);

            // Dispatch event to populate the `sales_sequence_meta` tables and such:
            $this->eventManager->dispatch('store_add', ['store' => $store]);
        }
    }
}
```

It is the `$this->eventManager->dispatch()` that triggers the following stack trace:

```
Installing data...

  [Elasticsearch\Common\Exceptions\BadRequest400Exception]
  {"error":{"root_cause":[{"type":"index_already_exists_exception","reason":"index [magento2_de_catalog_product_20180919_124316/w_hP07ExSmK7V898l_NHeQ] already exists","index_uuid":"w_hP07ExSmK7V898l_NHeQ","index":"magento2_de_catalog_product_20180919_124316"}],"type":
  "index_already_exists_exception","reason":"index [magento2_de_catalog_product_20180919_124316/w_hP07ExSmK7V898l_NHeQ] already exists","index_uuid":"w_hP07ExSmK7V898l_NHeQ","index":"magento2_de_catalog_product_20180919_124316"},"status":400}


Exception trace:
 () at /var/www/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:610
 Elasticsearch\Connections\Connection->process4xxError() at /var/www/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:273
 Elasticsearch\Connections\Connection->Elasticsearch\Connections\{closure}() at /var/www/vendor/react/promise/src/FulfilledPromise.php:25
 React\Promise\FulfilledPromise->then() at /var/www/vendor/guzzlehttp/ringphp/src/Future/CompletedFutureValue.php:55
 GuzzleHttp\Ring\Future\CompletedFutureValue->then() at /var/www/vendor/guzzlehttp/ringphp/src/Core.php:341
 GuzzleHttp\Ring\Core::proxy() at /var/www/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:294
 Elasticsearch\Connections\Connection->Elasticsearch\Connections\{closure}() at /var/www/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:171
 Elasticsearch\Connections\Connection->performRequest() at /var/www/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Transport.php:106
 Elasticsearch\Transport->performRequest() at /var/www/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Namespaces/AbstractNamespace.php:72
 Elasticsearch\Namespaces\AbstractNamespace->performRequest() at /var/www/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Namespaces/IndicesNamespace.php:718
 Elasticsearch\Namespaces\IndicesNamespace->create() at /var/www/vendor/smile/elasticsuite/src/module-elasticsuite-core/Client/Client.php:68
 Smile\ElasticsuiteCore\Client\Client->createIndex() at /var/www/vendor/smile/elasticsuite/src/module-elasticsuite-core/Index/IndexOperation.php:136
 Smile\ElasticsuiteCore\Index\IndexOperation->createIndex() at /var/www/vendor/smile/elasticsuite/src/module-elasticsuite-core/Indexer/GenericIndexerHandler.php:147
 Smile\ElasticsuiteCore\Indexer\GenericIndexerHandler->cleanIndex() at /var/www/vendor/magento/module-catalog-search/Model/Indexer/Fulltext/Store.php:58
 Magento\CatalogSearch\Model\Indexer\Fulltext\Store->clearIndex() at /var/www/vendor/magento/module-catalog-search/Model/Indexer/Fulltext/Store.php:69
 Magento\CatalogSearch\Model\Indexer\Fulltext\Store->execute() at /var/www/vendor/magento/framework/Event/Invoker/InvokerDefault.php:72
 Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod() at /var/www/vendor/magento/framework/Event/Invoker/InvokerDefault.php:60
 Magento\Framework\Event\Invoker\InvokerDefault->dispatch() at /var/www/vendor/magento/framework/Event/Manager.php:66
 Magento\Framework\Event\Manager->dispatch() at /var/www/generated/code/Magento/Framework/Event/Manager/Proxy.php:95
 Magento\Framework\Event\Manager\Proxy->dispatch() at /var/www/app/code/Vendor/Module/Setup/InstallData.php:162
```

### Expected result

It should install (it always did).

### Actual result

It throws the following error:

```
  [Elasticsearch\Common\Exceptions\BadRequest400Exception]
  {"error":{"root_cause":[{"type":"index_already_exists_exception","reason":"index [magento2_de_catalog_product_20180919_124316/w_hP07ExSmK7V898l_NHeQ] already exists","index_uuid":"w_hP07ExSmK7V898l_NHeQ","index":"magento2_de_catalog_product_20180919_124316"}],"type":
  "index_already_exists_exception","reason":"index [magento2_de_catalog_product_20180919_124316/w_hP07ExSmK7V898l_NHeQ] already exists","index_uuid":"w_hP07ExSmK7V898l_NHeQ","index":"magento2_de_catalog_product_20180919_124316"},"status":400}
```

**Comment by kanduvisla** [comment]
Update: I also triede it using ElasticSearch 6.4.0 (given that the module aims to be 6.x-compatible), but I still get the error, only a slightly different message:

```
  [Elasticsearch\Common\Exceptions\BadRequest400Exception]
  {"error":{"root_cause":[{"type":"resource_already_exists_exception","reason":"index [magento2_de_catalog_product_20180919_125659/3WRpH1cjS3KoDm4swDanxw] already exists","index_uuid":"3WRpH1cjS3KoDm4swDanxw","index":"magento2_de_catalog_product_20180919_125659"}],"typ
  e":"resource_already_exists_exception","reason":"index [magento2_de_catalog_product_20180919_125659/3WRpH1cjS3KoDm4swDanxw] already exists","index_uuid":"3WRpH1cjS3KoDm4swDanxw","index":"magento2_de_catalog_product_20180919_125659"},"status":400}
```

So it's now `resource_already_exists_exception` instead of `index_already_exists_exception`

**Comment by Streetec** [comment]
Based on your last error report:
Everything i see are the keywords `resource_already_exists_exception`.
Do you run the Elasticserver on the same Port and Index name as the one on 2.2.5 ?

Maybe try to change the Index names and port or delete the index based on the official elasticsearch wiki: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-delete-index.html

**Comment by kanduvisla** [comment]
I've looked with Kabana and there are no existing indices. The name is every time unique (according to the timestamp), so it's strange that this error is thrown on a fresh install.

**Comment by kanduvisla** [comment]
Hm, I think I found the root of my problem:

If simply put an `echo`-statement in `\Smile\ElasticsuiteCore\Index\IndexOperation::createIndex()` to see when indexes are created and with what name:

```
[Progress: 426 / 617]
Module 'Vendor_Store':
Installing data...
CREATE INDEX magento2_de_catalog_product_20180920_093801
CREATE INDEX magento2_de_catalog_product_20180920_093801
```
As you can see in my install-script, I create several store views in a `foreach`-loop, so the name will be identical, triggering the error. This caused me to search for the reason why this didn't throw an error before. I could find any specific changes in the ElasticSuite code here, the indexers were already registered in the older version (2.5.13) and the method that created the name also hadn't changed.

So I decided to rollback my 2.2.5 update and try again... Here it works, but I noticed that the indexes had the proper store codes (`de`, `nl`, `en`, etc.). I noticed that the test results first only had `de`:

```
CREATING INDEX: magento2_nl_catalog_product_20180920_095911
CREATING INDEX: magento2_de_catalog_product_20180920_095911
CREATING INDEX: magento2_en_catalog_product_20180920_095911
CREATING INDEX: magento2_fr_catalog_product_20180920_095911
```

Back to 2.2.6:

I added an `echo`-statement in my install script:

```
CREATING STORE WITH CODE de
CREATING INDEX: magento2_de_catalog_product_20180920_104007
CREATING INDEX: magento2_de_catalog_product_20180920_104007
```

Hm, it seems that creating a store in my install script triggers the indexer twice. Time to get some stack traces. It seems that the following dispatched event also triggers a reindex:

    // Dispatch event to populate the `sales_sequence_meta` tables and such:
    $this->eventManager->dispatch('store_add', ['store' => $store]);

This is in my install script to populate the `sales_sequence_meta`-tables (this event is registered for this functionality in `Magento_Sales/etc/events.xml`). The event is also caught by `Magento_CatalogSearch/etc/events.xml` to trigger the reindex). So if the event is dispatched once, everything works fine, but what if it gets dispatched twice? Time to diff Magento 2.2.5 with 2.2.6:

https://github.com/magento/magento2/commit/55e23bced838e97f5a903bed0efb2c7951abbed3#diff-041ce05c99a40e932cc0eb08cad3e269R1057

In this commit we see that `Magento\Store\Model\Store.php::afterSave()` is changed to dispatch an additional `store_add`-event after saving. That's the cause of this issue now. 

So in my case the fix is simply to remove my line to dispatch the event manually (since in Magento 2.2.6 it's dispatched on a regular save). I've did that and now the CLI install works perfectly.

So far my research.

**As far as ElasticSuite is concerned: what is now to proper action to take?** In theory it could always happen that for some reasons the reindex-command is triggered in the same second. Is the current implementation still accurate? Should the name generation of the indexer be changed? For example, add a random hash or something after the timestamp? This would ensure that there is always an unique name for the index, how silly it might seem that the same index gets reindexed in the same second. Or perhaps throw a more meaningful exception that doesn't cost a develop a couple of hours to trace to what's the actual cause?

I'm really curious about the ElasticSuite developers opinion on this one.

**Comment by romainruaud** [comment]
Wow !

such a great investigation. I'm always pleased to read issues like this one, since it's really complete about what's happening inside Magento's core and save us a lot of time dealing with it.

So first of all, thank you for the time you spent on this one.

About triggering the reindex on the same second : it _should not_ happen, because Magento has a locking mechanism built-in that prevent from running more than one reindex simultaneously.

I agree that adding more robust naming could be an option to secure more this part. But the "indice pattern" is already customizable via the configuration, actually it's `{{YYYYMMdd}}_{{HHmmss}}`.

Actually I'm not sure if we'd be able to be more precise with horodating, everything relies here : https://github.com/Smile-SA/elasticsuite/blob/master/src/module-elasticsuite-core/Helper/IndexSettings.php#L82

Maybe if you have some time you could do some testing here with other horodating patterns (I'm not sure what is the max-depth we can extract from a Zend_Date actually).

That being said, I consider this one as low-priority for now, because of the locking mechanism, and the fact that your initial issue was due to Magento core changes.

Best regards, and thank you again for this detailed issue !

**Comment by convenient** [comment]
I'm getting this kind of error when running integration tests on 2.2.6 (which means also a fresh install).

Thanks for your debugging notes. I hope to get further into this issue.

**Comment by dharake** [comment]
Hi guys, 

I can chime in here with a similar scenario to you.

I've got a M2.3 CE in development, running LEMP w/ Varnish, and elastic-server on a seperate VM. 

All seemed to be running fine for a while, then today I start getting the following in my exception.log. Like @kanduvisla I'm running multiple store fronts.

`[2019-03-04 08:59:08] main.CRITICAL: {"error":{"root_cause":[{"type":"resource_already_exists_exception","reason":"index [magento2_at_catalog_product_20190304_085908/0KTWjsdjSU-CpXcWlbZlIA] already exists","index_uuid":"0KTWjsdjSU-CpXcWlbZlIA","index":"magento2_at_catalog_product_20190304_085908"}],"type":"resource_already_exists_exception","reason":"index [magento2_at_catalog_product_20190304_085908/0KTWjsdjSU-CpXcWlbZlIA] already exists","index_uuid":"0KTWjsdjSU-CpXcWlbZlIA","index":"magento2_at_catalog_product_20190304_085908"},"status":400} {"exception":"[object] (Elasticsearch\\Common\\Exceptions\\BadRequest400Exception(code: 400): {\"error\":{\"root_cause\":[{\"type\":\"resource_already_exists_exception\",\"reason\":\"index [magento2_at_catalog_product_20190304_085908/0KTWjsdjSU-CpXcWlbZlIA] already exists\",\"index_uuid\":\"0KTWjsdjSU-CpXcWlbZlIA\",\"index\":\"magento2_at_catalog_product_20190304_085908\"}],\"type\":\"resource_already_exists_exception\",\"reason\":\"index [magento2_at_catalog_product_20190304_085908/0KTWjsdjSU-CpXcWlbZlIA] already exists\",\"index_uuid\":\"0KTWjsdjSU-CpXcWlbZlIA\",\"index\":\"magento2_at_catalog_product_20190304_085908\"},\"status\":400} at /var/www/html/magento/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:610)"} []



[2019-03-04 09:07:11] main.CRITICAL: {"error":{"root_cause":[{"type":"resource_already_exists_exception","reason":"index [magento2_pt_catalog_product_20190304_090711/yPVyzQURSLm8ssqlgBGQVw] already exists","index_uuid":"yPVyzQURSLm8ssqlgBGQVw","index":"magento2_pt_catalog_product_20190304_090711"}],"type":"resource_already_exists_exception","reason":"index [magento2_pt_catalog_product_20190304_090711/yPVyzQURSLm8ssqlgBGQVw] already exists","index_uuid":"yPVyzQURSLm8ssqlgBGQVw","index":"magento2_pt_catalog_product_20190304_090711"},"status":400} {"exception":"[object] (Elasticsearch\\Common\\Exceptions\\BadRequest400Exception(code: 400): {\"error\":{\"root_cause\":[{\"type\":\"resource_already_exists_exception\",\"reason\":\"index [magento2_pt_catalog_product_20190304_090711/yPVyzQURSLm8ssqlgBGQVw] already exists\",\"index_uuid\":\"yPVyzQURSLm8ssqlgBGQVw\",\"index\":\"magento2_pt_catalog_product_20190304_090711\"}],\"type\":\"resource_already_exists_exception\",\"reason\":\"index [magento2_pt_catalog_product_20190304_090711/yPVyzQURSLm8ssqlgBGQVw] already exists\",\"index_uuid\":\"yPVyzQURSLm8ssqlgBGQVw\",\"index\":\"magento2_pt_catalog_product_20190304_090711\"},\"status\":400} at /var/www/html/magento/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:610)"} []



[2019-03-04 09:58:49] main.CRITICAL: {"error":{"root_cause":[{"type":"resource_already_exists_exception","reason":"index [magento2_fr_catalog_product_20190304_095849/nqQG8xvaQmGG73vtjqlq1w] already exists","index_uuid":"nqQG8xvaQmGG73vtjqlq1w","index":"magento2_fr_catalog_product_20190304_095849"}],"type":"resource_already_exists_exception","reason":"index [magento2_fr_catalog_product_20190304_095849/nqQG8xvaQmGG73vtjqlq1w] already exists","index_uuid":"nqQG8xvaQmGG73vtjqlq1w","index":"magento2_fr_catalog_product_20190304_095849"},"status":400} {"exception":"[object] (Elasticsearch\\Common\\Exceptions\\BadRequest400Exception(code: 400): {\"error\":{\"root_cause\":[{\"type\":\"resource_already_exists_exception\",\"reason\":\"index [magento2_fr_catalog_product_20190304_095849/nqQG8xvaQmGG73vtjqlq1w] already exists\",\"index_uuid\":\"nqQG8xvaQmGG73vtjqlq1w\",\"index\":\"magento2_fr_catalog_product_20190304_095849\"}],\"type\":\"resource_already_exists_exception\",\"reason\":\"index [magento2_fr_catalog_product_20190304_095849/nqQG8xvaQmGG73vtjqlq1w] already exists\",\"index_uuid\":\"nqQG8xvaQmGG73vtjqlq1w\",\"index\":\"magento2_fr_catalog_product_20190304_095849\"},\"status\":400} at /var/www/html/magento/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:610)"} []
`
I'm wondering if enabling RabbitMQ can help in this situation?

**Comment by philippsander** [comment]
This issue is really annoying for us unfortunately

We are setting up the projects quite often to execute the unit-tests.
So if someone would try to solve this I would be super happy

**Comment by romainruaud** [comment]
I'm closing this very old issue.

Feel free to open a new one if needed.

Regards

**Comment by mfickers** [comment]
I've encountered this issue during Magento installation in the integration test job of our pipeline.

One of our modules includes a data patch that creates some categories. When these categories are saved, the category fulltext index is updated. Since the index does not exist, it will be created.

Then the install scripts for the Elasticsuite modules are run, which include \Smile\ElasticsuiteCatalog\Setup\InstallData. This runs a reindexAll on the catagory fulltext index. This will always create a new index.

If both data patch and install script are run in the same second, then the index name will be the exact same and result in a `resource_already_exists_exception`.

For version 2.11 there is a workaround. Because Zend_Date has been replaced with PHP native DateTime in #2832, it is now possible to change the `smile_elasticsuite_core_base_settings/indices_settings/indices_pattern` to `{{Ymd}}_{{His}}_{{v}}` and the index names will be suffixed with the milliseconds.
